### PR TITLE
DeepSpeed Monitor Module

### PR DESCRIPTION
**Pull request by Quentin-Anthony**
This PR introduces a general `Monitor` tool which manages logging with torch's TensorBoard, [WandB](https://wandb.ai/home), or CSV files. 

This would allow users to add custom logging such as https://github.com/microsoft/DeepSpeed/issues/327, and enables DeepSpeed to easily apply logging outside the engine.

TODO:

- [x] Basic functionality for each monitor
- [x] Refactor into module
- [x] Add documentation + examples
- [x] Add pydantic config support
- [x] Add monitor tests

### CODE REVIEW

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -488,53 +490,8 @@ def curriculum_enabled(self):
...
-        except ImportError:
-            print(
-                'If you want to use tensorboard logging please `pip install tensorboard`'
-            )
-            raise
-        from torch.utils.tensorboard import SummaryWriter
-
-        return SummaryWriter(log_dir=log_dir)
+    def monitor_enabled(self):
+        return self._config.tensorboard_enabled or self._config.wandb_enabled or self._config.csv_monitor_enabled
```
Comment (jeffra) [inline deepspeed/runtime/engine.py:494]: Can we move this check to `utils/monitor.py` somewhere? This way in the future if we add new monitoring backends we don't have to add any code to the engine and only to the monitor itself.

Comment (Quentin-Anthony) [reply deepspeed/runtime/engine.py]: Done

File: deepspeed/monitor/wandb.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from .utils import *
+from .monitor import *
```
Comment (jeffra) [inline deepspeed/monitor/wandb.py:2]: Can we use explicit imports here? I think there are only two things we need to import right? `check_wandb_availability` and `Monitor`?

Comment (Quentin-Anthony) [reply deepspeed/monitor/wandb.py]: Done

**Comment by rocm-mici** [comment]
Can one of the admins verify this patch?

**Comment by Quentin-Anthony** [comment]
@jeffra -- This is ready to review again when you get time

**Review by jeffra — APPROVED** [review]
🚀

**Pull request merged**